b2sdk/utils/curl.py

```diff
@@ -0,0 +1,431 @@
+            status, _count = self.multi.perform()
+            # Usual culprits here could be "out of memory" and "internal error",
+            # we can't really do anything about either of them.
+            assert status == pycurl.E_MULTI_OK, status
+
+            # Check whether any of the curls finished, in one way or another.
+            remaining_messages_count, successful_curls, failed_curls = self.multi.info_read()
+            # Remove each finished curl from our multi object and close it.
+            finished_curls = successful_curls + [elem[0] for elem in failed_curls]
+            # TODO: additional handling for failed curls.
+            for curl in finished_curls:
+                self.multi.remove_handle(curl)
+                streamer = self.mapping[curl]
+                streamer.close()
+
+
+@dataclass
+class CurlResponse:
+    """
+    Analog of requests.Response
+
+    Provides basic interface for interaction with HTTP.
+
+    Note that in streaming mode it can be completely "empty",
+    that is neither status code nor headers are downloaded initially.
+    This means that first call to `status_code`, `headers` or
+    `iter_content` (through e.g. `content`) will start the actual operations.
+    """
+
+    streamer: CurlStreamer
+    request: CurlRequest
+
+    content_cache: Optional[bytes] = None
+
+    @property
+    def url(self) -> str:
+        return self.request.url
+
+    @property
+    def status_code(self) -> int:
+        return self.streamer.status_code
+
+    @property
+    def headers(self) -> CaseInsensitiveDict:
+        return self.streamer.headers
+
+    @property
+    def ok(self) -> bool:
+        return 200 <= self.status_code < 400
+
+    @property
+    def content(self) -> bytes:
+        if self.content_cache is None:
+            self.content_cache = b''.join(self.iter_content(1024))
+        return self.content_cache
+
+    def iter_content(self, chunk_size: int) -> Iterator[bytes]:
+        while True:
+            data = self.streamer.read(chunk_size)
+            if not data:
+                if self.streamer.is_reading_done:
+                    break
+                else:
+                    continue
+
+            yield data
+
+    def close(self):
+        pass
+
+
+def headers_to_list(headers: Dict[str, str]) -> List[str]:
+    result = []
+    for key, value in headers.items():
+        if value is None:
+            continue
+
+        if isinstance(value, bytes):
+            value = value.decode('utf8')
+
+        result.append(f'{key}: {value}')
+
+    return result
+
+
+class CurlSession:
+    TIMEOUT = 5
+    EXPECT_100_TIMEOUT = 10
+    BUFFER_SIZE_BYTES = 10 * 1024 * 1024
+    VERBOSE = False
+    NO_SIGNAL = 0
+
+    def __init__(self, *args, **kwargs):
+        self.adapters = CurlAdapters()
+        self.manager = CurlManager()
+
+    def mount(self, scheme, adapter):
+        self.adapters.adapters[scheme] = adapter
+
+    def request(
```

Comment:
maybe it should be obvious, but I can't find any line explaining it: does this `session.request` work as `requests.session.request` i.e. SSL handshakes are only performed on the first requests and cookies are persisted?

Reply:
Thank you for bringing this up.

I've checked cookies first, and no. Cookies were disabled to the level that they would not even persist through redirection. For these, I had to check the implementation and it seems that the only thing we can do, is to use `libcurl` default. And that's "writing to a file" and "reading from a file". And this file is specified as a string, so no `BytesIO` for us.

As for SSL, curls are not re-used, so unless some magical trickery happens underneath (I'll check), I'd say – no.

Reply:
I'll most probably go with `http.cookiejar` and manual `cookie` handling then.

Reply:
Cookies are handled now.

Reply:
As for SSL – I've added curl handles caching. Whenever handle is reused, it should reuse the connection (so SSL should already be established).
In streamed mode, pipelinening and multiplexing is used (alongside cache) to ensure that, whenever viable, queries will reuse connections.